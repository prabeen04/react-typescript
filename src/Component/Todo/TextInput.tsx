import * as React from 'react'

interface ITextInputProps {
    value: string;
    name: string;
    onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
    placeholder?: string;
    className?: string
    style?: object
}
export default function TextInput(props: ITextInputProps) {
    const { value, placeholder, onChange, ...rest } = props;
    return (
        <>
            <input
                name={name}
                value={value}
                placeholder={placeholder}
                onChange={onChange}
                {...rest}
            />
        </>
    )
}
