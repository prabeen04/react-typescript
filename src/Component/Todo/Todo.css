.todo-wrapper{
}
.todo-list-group{
    padding: 0.5rem;
}
.todo-list-group-item{
    padding: 0.9rem;
    margin: 0.5rem 0.3rem;
    width: auto;
    border-left: 4px solid orangered;
    background: rgba(51, 51, 51, 0.13);
    border-radius: 0 .5rem 0.5rem 0.5rem 0;
}
.todo-view{
    display: flex;
    justify-content: space-between;
}
.todo-edit{
    display: flex;
    align-items: center;
    justify-content: space-between;
}