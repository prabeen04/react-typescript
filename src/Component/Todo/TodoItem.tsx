import * as React from 'react'
import { ITodo } from "./Todo";
import ViewEdit from "../../Hooks/ViewEdit";
import TextInput from "./TextInput";
export default function TodoItem(
    { todo, deleteTodo, toggleTodo, updateTodo }
        : {
            todo: ITodo,
            deleteTodo: (todo: ITodo) => void,
            toggleTodo: (todo: ITodo) => void,
            updateTodo: (todo: ITodo, index: string) => void,
        }): JSX.Element {
    const { viewType, toggleViewType } = ViewEdit()
    const [title, setTitle] = React.useState<string>(todo.title)
    return (
        <div className={`todo-list-group-item ${todo.isCompleted ? 'completed-todo' : ''}`}>
            {
                viewType === 'view'
                    ? <span className='todo-view'>
                        <span
                            onClick={() => toggleTodo(todo)}
                            style={{
                                fontSize: 24,
                                color: todo.isCompleted ? 'rgb(1, 102, 63)' : '#f23a21',
                                textDecoration: todo.isCompleted ? 'line-through' : 'none',
                            }}>
                            {todo.title}</span>
                        <span>
                            <button className='btn btn-warning yellow' onClick={() => toggleTodo(todo)}>
                            {todo.isCompleted ? 'Incomplete' : 'Completed'}
                            </button>&nbsp;
                            <button className='btn btn-secondary sky-blue' onClick={toggleViewType}>Edit</button>&nbsp;
                          <button className='btn btn-danger' onClick={() => deleteTodo(todo)}>Delete</button>
                        </span>
                    </span>
                    : <span className='todo-edit'>
                        <TextInput
                            className='dark-input'
                            name=''
                            onChange={({ target: { value } }) => setTitle(value)}
                            value={title}
                            style={{ width: 400 }}
                        />
                        <span>

                            <button
                                className='btn btn-primary'
                                onClick={() => {
                                    updateTodo(todo, title)
                                    toggleViewType()
                                }}>Update</button>&nbsp;
                            <button className='btn btn-danger' onClick={toggleViewType}>Cancel</button>
                        </span>
                    </span>
            }

        </div>
    )
}
