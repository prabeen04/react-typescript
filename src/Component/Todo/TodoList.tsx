import * as React from 'react'
import { ITodo } from './Todo'
import TodoItem from "./TodoItem";
export interface ITodoListProps {
    todos: ITodo[];
    deleteTodo: (todo: ITodo) => void;
    toggleTodo: (todo: ITodo) => void;
    updateTodo: (todo: ITodo, newTodo: string) => void;
}
export default function TodoList(props: ITodoListProps): JSX.Element {
    const { todos } = props
    return (
        <ul className='list-group list-group-flush' data-test='TodoList'>
            {
                !!todos && todos.map((todo: ITodo) => <TodoItem
                    key={todo.id}
                    todo={todo}
                    deleteTodo={props.deleteTodo}
                    toggleTodo={props.toggleTodo}
                    updateTodo={props.updateTodo}
                />)
            }
        </ul>
    )
}
