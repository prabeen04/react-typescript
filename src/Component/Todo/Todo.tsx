import * as React from 'react'
import TextInput from './TextInput';
import TodoList from './TodoList'
import './Todo.css';

export interface ITodo {
    title: string;
    isCompleted: boolean;
    id: number;
}
export type InputElem = React.ChangeEvent<HTMLInputElement>
export type FormElem = React.FormEvent<HTMLFormElement>

export default function Todo(): JSX.Element {
    const [todo, setTodo] = React.useState<string>('')
    const [todos, setTodos] = React.useState<ITodo[]>([])

    function handleChange(e: InputElem): void {
        setTodo(e.target.value)
    }
    function handleSubmit(e: FormElem): void {
        e.preventDefault()
        let newTodos = [...todos, { title: todo, isCompleted: false, id: Date.now() }]
        setTodos(newTodos)
        setTodo('')
    }
    function deleteTodo(todo: ITodo): void {
        let newTodos = [...todos]
        let newArr = newTodos.filter((singleTodo: ITodo) => singleTodo.id !== todo.id)
        setTodos(newArr)
    }
    function toggleTodo(todo: ITodo): void {
        let newTodos = [...todos]
        let newArr = newTodos.map((singleTodo: ITodo) => {
            if (singleTodo.id === todo.id) {
                return { ...singleTodo, isCompleted: !singleTodo.isCompleted }
            } else {
                return singleTodo
            }
        })
        setTodos(newArr)
    }
    function updateTodo(todo: ITodo, newTodo: string): void {
        console.log('object')
        let newTodos = [...todos]
        let newArr = newTodos.map((singleTodo: ITodo) => {
            if (singleTodo.id === todo.id) {
                return { ...singleTodo, title: newTodo }
            } else {
                return singleTodo
            }
        })
        setTodos(newArr)
    }
    return (
        <>
            <div className='todo-wrapper'>
                <form onSubmit={handleSubmit} data-test='Todo'>
                    <br/>
                    <div className='todo-form'>
                        <TextInput
                            name='todo'
                            placeholder='Add task'
                            onChange={handleChange}
                            value={todo}
                            className='dark-input'
                            style={{ width: 400 }}
                        />&nbsp;&nbsp;
                    <button className='btn btn-success red' disabled={!todo}>
                    <i className="fas fa-paper-plane"></i>&nbsp;&nbsp;
                     Add</button>
                    </div>
                </form>
                <TodoList
                    todos={todos}
                    deleteTodo={deleteTodo}
                    toggleTodo={toggleTodo}
                    updateTodo={updateTodo}
                />
            </div>
        </>
    )
}

