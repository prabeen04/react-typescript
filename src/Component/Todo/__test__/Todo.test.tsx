import * as React from 'react';
import { shallow } from "enzyme";
import Todo from "../Todo";

describe('<Todo />', () => {
    let wrapper: any;
    beforeEach(() => {
        wrapper = shallow(<Todo />)
    })

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('should render without crashing', () => {
        expect(wrapper.find('[data-test="Todo"]').length).toEqual(1)
    })
})