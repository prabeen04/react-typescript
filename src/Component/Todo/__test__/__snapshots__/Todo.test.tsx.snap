// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo /> should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todo />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="todo-wrapper"
      >
        <form
          data-test="Todo"
          onSubmit={[Function]}
        >
          <br />
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "center",
              }
            }
          >
            <TextInput
              className="dark-input"
              name="todo"
              onChange={[Function]}
              placeholder="Enter todo"
              style={
                Object {
                  "width": 400,
                }
              }
              value=""
            />
              
            <button
              className="btn btn-success"
              disabled={true}
            >
              Add
            </button>
          </div>
        </form>
        <TodoList
          deleteTodo={[Function]}
          todos={Array []}
          toggleTodo={[Function]}
          updateTodo={[Function]}
        />
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            data-test="Todo"
            onSubmit={[Function]}
          >
            <br />
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "justifyContent": "center",
                }
              }
            >
              <TextInput
                className="dark-input"
                name="todo"
                onChange={[Function]}
                placeholder="Enter todo"
                style={
                  Object {
                    "width": 400,
                  }
                }
                value=""
              />
                
              <button
                className="btn btn-success"
                disabled={true}
              >
                Add
              </button>
            </div>
          </form>,
          <TodoList
            deleteTodo={[Function]}
            todos={Array []}
            toggleTodo={[Function]}
            updateTodo={[Function]}
          />,
        ],
        "className": "todo-wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <br />,
              <div
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "justifyContent": "center",
                  }
                }
              >
                <TextInput
                  className="dark-input"
                  name="todo"
                  onChange={[Function]}
                  placeholder="Enter todo"
                  style={
                    Object {
                      "width": 400,
                    }
                  }
                  value=""
                />
                  
                <button
                  className="btn btn-success"
                  disabled={true}
                >
                  Add
                </button>
              </div>,
            ],
            "data-test": "Todo",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <TextInput
                    className="dark-input"
                    name="todo"
                    onChange={[Function]}
                    placeholder="Enter todo"
                    style={
                      Object {
                        "width": 400,
                      }
                    }
                    value=""
                  />,
                  "  ",
                  <button
                    className="btn btn-success"
                    disabled={true}
                  >
                    Add
                  </button>,
                ],
                "style": Object {
                  "alignItems": "center",
                  "display": "flex",
                  "justifyContent": "center",
                },
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "dark-input",
                    "name": "todo",
                    "onChange": [Function],
                    "placeholder": "Enter todo",
                    "style": Object {
                      "width": 400,
                    },
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "  ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Add",
                    "className": "btn btn-success",
                    "disabled": true,
                  },
                  "ref": null,
                  "rendered": "Add",
                  "type": "button",
                },
              ],
              "type": "div",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "deleteTodo": [Function],
            "todos": Array [],
            "toggleTodo": [Function],
            "updateTodo": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="todo-wrapper"
        >
          <form
            data-test="Todo"
            onSubmit={[Function]}
          >
            <br />
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "justifyContent": "center",
                }
              }
            >
              <TextInput
                className="dark-input"
                name="todo"
                onChange={[Function]}
                placeholder="Enter todo"
                style={
                  Object {
                    "width": 400,
                  }
                }
                value=""
              />
                
              <button
                className="btn btn-success"
                disabled={true}
              >
                Add
              </button>
            </div>
          </form>
          <TodoList
            deleteTodo={[Function]}
            todos={Array []}
            toggleTodo={[Function]}
            updateTodo={[Function]}
          />
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <form
              data-test="Todo"
              onSubmit={[Function]}
            >
              <br />
              <div
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "justifyContent": "center",
                  }
                }
              >
                <TextInput
                  className="dark-input"
                  name="todo"
                  onChange={[Function]}
                  placeholder="Enter todo"
                  style={
                    Object {
                      "width": 400,
                    }
                  }
                  value=""
                />
                  
                <button
                  className="btn btn-success"
                  disabled={true}
                >
                  Add
                </button>
              </div>
            </form>,
            <TodoList
              deleteTodo={[Function]}
              todos={Array []}
              toggleTodo={[Function]}
              updateTodo={[Function]}
            />,
          ],
          "className": "todo-wrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <br />,
                <div
                  style={
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "justifyContent": "center",
                    }
                  }
                >
                  <TextInput
                    className="dark-input"
                    name="todo"
                    onChange={[Function]}
                    placeholder="Enter todo"
                    style={
                      Object {
                        "width": 400,
                      }
                    }
                    value=""
                  />
                    
                  <button
                    className="btn btn-success"
                    disabled={true}
                  >
                    Add
                  </button>
                </div>,
              ],
              "data-test": "Todo",
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <TextInput
                      className="dark-input"
                      name="todo"
                      onChange={[Function]}
                      placeholder="Enter todo"
                      style={
                        Object {
                          "width": 400,
                        }
                      }
                      value=""
                    />,
                    "  ",
                    <button
                      className="btn btn-success"
                      disabled={true}
                    >
                      Add
                    </button>,
                  ],
                  "style": Object {
                    "alignItems": "center",
                    "display": "flex",
                    "justifyContent": "center",
                  },
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "className": "dark-input",
                      "name": "todo",
                      "onChange": [Function],
                      "placeholder": "Enter todo",
                      "style": Object {
                        "width": 400,
                      },
                      "value": "",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "  ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Add",
                      "className": "btn btn-success",
                      "disabled": true,
                    },
                    "ref": null,
                    "rendered": "Add",
                    "type": "button",
                  },
                ],
                "type": "div",
              },
            ],
            "type": "form",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "deleteTodo": [Function],
              "todos": Array [],
              "toggleTodo": [Function],
              "updateTodo": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
